data_structures/rb_tree2.cpp

```diff
@@ -0,0 +1,1703 @@
+                        // root replaced with a null pointer to empty tree
+                        root = nullptr;
+                        replacement_node = root;
+                    } else {
+                        // node's location is still needed for
+                        // fixing tree, so it's only marked as null for now
+                        node_to_delete->treatAsNull = true;
+                    }
+                } else {
+                    // node to delete has one child -- node's
+                    // child moves up to replace deleted node
+
+                    // removing a root with one child
+                    if (root == node_to_delete) {
+                        // root replaced with root's child and
+                        // root properties are reset
+                        root = MeToOnlyChildPtr(node_to_delete);
+                        root->parent = nullptr;
+                        root->color = COLOR_BLACK;
+                        replacement_node = root;
+                    } else {
+                        // setting parent and child pointers
+                        // for the node's one child
+                        ParentToMePtr(node_to_delete) =
+                        MeToOnlyChildPtr(node_to_delete);
+
+                        MeToOnlyChildPtr(node_to_delete)->parent =
+                        node_to_delete->parent;
+
+                        replacement_node = MeToOnlyChildPtr(node_to_delete);
+                    }
+                }
+                // Now check Rotate/Recolor conditions...
+                if (root != nullptr) {
+                    // a replacement node that's red or at the root
+                    // just needs to be recolored black
+                    if (replacement_node == root ||
+                    replacement_node->color == COLOR_RED) {
+                        replacement_node->color = COLOR_BLACK;
+                    } else if (node_to_delete->color == COLOR_BLACK &&
+                    (replacement_node->color == COLOR_BLACK ||
+                    replacement_node->treatAsNull == true)) {
+                        // deleted node was black and replacement node is
+                        // black/null, causes double-blackness problem in
+                        // tree that needs to be fixed
+                        replacement_node->color = COLOR_DOUBLE_BLACK;
+                        fix_double_black(replacement_node);
+                    }
+                }
+                // if node to delete is a leaf, disconnect it from its
+                // parent after treating it as null to fix any
+                // double-blackness issues in tree
+                if (node_to_delete->treatAsNull == true) {
+                    ParentToMePtr(node_to_delete) = nullptr;
+                }
+                // finally delete node's memory and decrement tree size
+                delete node_to_delete;
+                numItems--;
+            }
+        }
```

Comment:
Are you sure you need this many `if` statements?

```cpp
            if (node_to_delete->left != nullptr &&
            node_to_delete->right != nullptr) {
                int ios_val = InOrderSuccessor(node_to_delete)->val;
                Remove(ios_val);
                node_to_delete->val = ios_val;
                return
            }
            ... // no need for this else block you can write this directly into the function without the else block
            // the same can be applied to your other `if else` blocks.
```

I suggest looking up "guard clauses" and understanding why they help with this messy `if else` block 😄

Reply:
Oh ok, I'll look them up and work on applying them to my code. Thank you for the advice!